File: ServidorArgentum.NET/wskapiAO.vb
Option Strict Off
Option Explicit On

Imports System.Collections.Generic

Module wskapiAO
    ''' <summary>
    ''' Maps socket IDs to user indices
    ''' </summary>
    Private SocketToUserMap As New Dictionary(Of Integer, Integer)
    
    ''' <summary>
    ''' Gets the IP address of a connected client
    ''' </summary>
    Public Function GetAscIP(ByVal socketID As Integer) As String
        Try
            Dim userIdx As Integer = BuscaSlotSock(socketID)
            If userIdx > 0 Then
                Return UserList(userIdx).ip
            End If
        Catch
        End Try
        
        Return "0.0.0.0"
    End Function
    
    ''' <summary>
    ''' Handles a connection request from a client
    ''' </summary>
    Private Sub HandleConnectionReceived(socketID As Integer, clientIP As String)
        ' This runs on the main thread via the Windows.Forms event system
        Dim NewIndex As Integer = NextOpenUser()
        
        If NewIndex <= MaxUsers Then
            Call UserList(NewIndex).incomingData.ReadASCIIStringFixed(UserList(NewIndex).incomingData.length)
            Call UserList(NewIndex).outgoingData.ReadASCIIStringFixed(UserList(NewIndex).outgoingData.length)
            
            UserList(NewIndex).ip = clientIP
            
            For i As Integer = 1 To BanIps.Count()
                Dim bannedIP As String = CStr(BanIps.Item(i))
                If bannedIP = clientIP Then
                    Call WriteErrorMsg(NewIndex, "Su IP se encuentra bloqueada en este servidor.")
                    Call FlushBuffer(NewIndex)
                    Call SocketManager.CloseSocket(socketID)
                    Return
                End If
            Next
            
            If NewIndex > LastUser Then
                LastUser = NewIndex
            End If
            
            UserList(NewIndex).ConnID = socketID
            UserList(NewIndex).ConnIDValida = True
            
            AgregaSlotSock(socketID, NewIndex)
        Else
            Dim errorMsg As String = Protocol.PrepareMessageErrorMsg("El servidor se encuentra lleno en este momento. Disculpe las molestias ocasionadas.")
            Dim data() As Byte = SocketManager.StringToBytes(errorMsg)
            SocketManager.SendData(socketID, data)
            SocketManager.CloseSocket(socketID)
        End If
    End Sub
    
    ''' <summary>
    ''' Handles data received from a client
    ''' </summary>
    Private Sub HandleDataReceived(socketID As Integer, data() As Byte)
        ' This now runs on the main thread via the synchronized queue system
        Dim userIndex As Integer = BuscaSlotSock(socketID)
        
        If userIndex > 0 Then
            With UserList(userIndex)
                Call .incomingData.WriteBlock(data)
                
                If .ConnID <> -1 Then
                    Call HandleIncomingData(userIndex)
                End If
            End With
        End If
    End Sub

    ''' <summary>
    ''' Handles closure of a socket connection
    ''' </summary>
    Private Sub HandleConnectionClosed(socketID As Integer)
        ' This runs on the main thread via the Windows.Forms event system
        Dim userIndex As Integer = BuscaSlotSock(socketID)

        If userIndex > 0 Then
            Call BorraSlotSock(socketID)

            UserList(userIndex).ConnID = -1
            UserList(userIndex).ConnIDValida = False

            Call EventoSockClose(userIndex)
        End If
    End Sub

    ''' <summary>
    ''' Handles connection error
    ''' </summary>
    Private Sub HandleConnectionError(errorMessage As String)
        ' This runs on the main thread via the Windows.Forms event system
        Debug.Print("Connection Error: " & errorMessage)
    End Sub
    
    ''' <summary>
    ''' Closes a socket connection
    ''' </summary>
    Public Function Winsock_Close(ByRef socketID As Integer) As Object
        ' This is called from the main thread
        Dim userIndex As Integer = BuscaSlotSock(socketID)

        If socketID > 0 Then
            SocketManager.CloseSocket(socketID)
        End If

        If userIndex > 0 Then
            Call BorraSlotSock(socketID)
            UserList(userIndex).ConnID = -1
            UserList(userIndex).ConnIDValida = False
            Call EventoSockClose(userIndex)
        End If

        Return Nothing
    End Function

    ''' <summary>
    ''' Initializes the socket API
    ''' </summary>
    Public Sub IniciaWsApi(ByVal port As Integer)
        ' Set up event handlers to process messages on the main thread
        AddHandler SocketManager.ConnectionReceived, AddressOf HandleConnectionReceived
        AddHandler SocketManager.DataReceived, AddressOf HandleDataReceived
        AddHandler SocketManager.ConnectionClosed, AddressOf HandleConnectionClosed
        AddHandler SocketManager.ServerError, AddressOf HandleConnectionError

        SocketManager.Initialize(port)
    End Sub
    
    ''' <summary>
    ''' Cleans up the socket API
    ''' </summary>
    Public Sub LimpiaWsApi()
        RemoveHandler SocketManager.ConnectionReceived, AddressOf HandleConnectionReceived
        RemoveHandler SocketManager.DataReceived, AddressOf HandleDataReceived
        RemoveHandler SocketManager.ConnectionClosed, AddressOf HandleConnectionClosed
        RemoveHandler SocketManager.ServerError, AddressOf HandleConnectionError

        SocketManager.Shutdown()
        
        SocketToUserMap.Clear()
    End Sub
    
    ''' <summary>
    ''' Sends data through a user's socket connection
    ''' </summary>
    Public Function WsApiEnviar(ByVal userIndex As Integer, ByRef message As String) As Integer
        On Error GoTo ErrorHandler
        
        Dim returnCode As Integer = 0
        
        Dim socketID As Integer = UserList(userIndex).ConnID
        
        If socketID <> -1 AndAlso UserList(userIndex).ConnIDValida Then
            Dim data() As Byte = SocketManager.StringToBytes(message)
            
            If Not SocketManager.SendData(socketID, data) Then
                returnCode = -1
            End If
            
            ' Process any pending messages that may have arrived
            SocketManager.ProcessPendingMessages()
        ElseIf socketID <> -1 AndAlso Not UserList(userIndex).ConnIDValida Then
            If Not UserList(userIndex).Counters.Saliendo Then
                returnCode = -1
            End If
        End If
        
        WsApiEnviar = returnCode
        
        Exit Function
        
ErrorHandler:
        Call UserList(userIndex).outgoingData.WriteASCIIStringFixed(message)
        Resume Next
    End Function
    
    ''' <summary>
    ''' Process any pending messages in the queue (can be called from main game loop)
    ''' </summary>
    Public Sub ProcessNetworkMessages()
        ' Call this from your main game loop or timer to ensure messages get processed
        SocketManager.ProcessPendingMessages()
    End Sub
    
    ''' <summary>
    ''' Reinitializes all sockets
    ''' </summary>
    Public Sub WSApiReiniciarSockets()
        Dim i As Integer
        
        For i = 1 To MaxUsers
            If UserList(i).ConnID <> -1 AndAlso UserList(i).ConnIDValida Then
                Call CloseSocket(i)
            End If
        Next i
        
        For i = 1 To MaxUsers
            UserList(i).incomingData = Nothing
            UserList(i).outgoingData = Nothing
        Next i
        
        ReDim UserList(MaxUsers)
        ArrayInitializers.InitializeStruct(UserList)
        For i = 1 To MaxUsers
            UserList(i).ConnID = -1
            UserList(i).ConnIDValida = False
            
            UserList(i).incomingData = New clsByteQueue
            UserList(i).outgoingData = New clsByteQueue
        Next i
        
        LastUser = 1
        NumUsers = 0
        
        SocketToUserMap.Clear()
        
        Call IniciaWsApi(Puerto)
    End Sub
    
    ''' <summary>
    ''' Finds the user index associated with a socket ID
    ''' </summary>
    Public Function BuscaSlotSock(ByVal socketID As Integer) As Integer
        Dim userIndex As Integer = -1
        
        If SocketToUserMap.TryGetValue(socketID, userIndex) Then
            Return userIndex
        End If
        
        Return -1
    End Function
    
    ''' <summary>
    ''' Associates a socket ID with a user index
    ''' </summary>
    Public Sub AgregaSlotSock(ByVal socketID As Integer, ByVal userIndex As Integer)
        Debug.Print("AgregaSlotSock: Socket " & socketID & " -> User " & userIndex)
        
        If SocketToUserMap.Count > MaxUsers Then
            Call CloseSocket(userIndex)
            Exit Sub
        End If
        
        SocketToUserMap(socketID) = userIndex
    End Sub
    
    ''' <summary>
    ''' Removes the association between a socket ID and a user index
    ''' </summary>
    Public Sub BorraSlotSock(ByVal socketID As Integer)
        Dim count As Integer = SocketToUserMap.Count
        
        SocketToUserMap.Remove(socketID)
        
        Debug.Print("BorraSlotSock: " & count & " -> " & SocketToUserMap.Count)
    End Sub
    
    ''' <summary>
    ''' Handles the event of a socket being closed
    ''' </summary>
    Public Sub EventoSockClose(ByVal userIndex As Integer)
        If Centinela.RevisandoUserIndex = userIndex Then 
            Call modCentinela.CentinelaUserLogout()
        End If
        
        If UserList(userIndex).flags.UserLogged Then
            Call CloseSocketSL(userIndex)
            Call Cerrar_Usuario(userIndex)
        Else
            Call CloseSocket(userIndex)
        End If
    End Sub
End Module

File: Legacy/Source/wskapiAO.cs
using System;
using System.Collections.Generic;

namespace Legacy;

internal static class wskapiAO
{
    /// <summary>
    ///     Maps socket IDs to user indices
    /// </summary>
    private static readonly Dictionary<int, int> SocketToUserMap = new();

    /// <summary>
    ///     Gets the IP address of a connected client
    /// </summary>
    public static string GetAscIP(int socketID)
    {
        try
        {
            var userIdx = BuscaSlotSock(socketID);
            if (userIdx > 0) return Declaraciones.UserList[userIdx].ip;
        }
        catch
        {
        }

        return "0.0.0.0";
    }

    /// <summary>
    ///     Handles a connection request from a client
    /// </summary>
    private static void HandleConnectionReceived(int socketID, string clientIP)
    {
        // This runs on the main thread via the Windows.Forms event system
        int NewIndex = UsUaRiOs.NextOpenUser();

        if (NewIndex <= Declaraciones.MaxUsers)
        {
            Declaraciones.UserList[NewIndex].incomingData
                .ReadASCIIStringFixed(Declaraciones.UserList[NewIndex].incomingData.length);
            Declaraciones.UserList[NewIndex].outgoingData
                .ReadASCIIStringFixed(Declaraciones.UserList[NewIndex].outgoingData.length);

            Declaraciones.UserList[NewIndex].ip = clientIP;

            for (int i = 0, loopTo = Declaraciones.BanIps.Count - 1; i <= loopTo; i++)
            {
                var bannedIP = Declaraciones.BanIps[i];
                if ((bannedIP ?? "") == (clientIP ?? ""))
                {
                    Protocol.WriteErrorMsg(Convert.ToInt16(NewIndex), "Su IP se encuentra bloqueada en este servidor.");
                    Protocol.FlushBuffer(Convert.ToInt16(NewIndex));
                    SocketManager.CloseSocket(socketID);
                    return;
                }
            }

            if (NewIndex > Declaraciones.LastUser) Declaraciones.LastUser = Convert.ToInt16(NewIndex);

            Declaraciones.UserList[NewIndex].ConnID = Convert.ToInt16(socketID);
            Declaraciones.UserList[NewIndex].ConnIDValida = true;

            AgregaSlotSock(socketID, NewIndex);
        }
        else
        {
            var errorMsg =
                Protocol.PrepareMessageErrorMsg(
                    "El servidor se encuentra lleno en este momento. Disculpe las molestias ocasionadas.");
            var data = SocketManager.StringToBytes(errorMsg);
            SocketManager.SendData(socketID, data);
            SocketManager.CloseSocket(socketID);
        }
    }

    /// <summary>
    ///     Handles data received from a client
    /// </summary>
    private static void HandleDataReceived(int socketID, byte[] data)
    {
        // This now runs on the main thread via the synchronized queue system
        var userIndex = BuscaSlotSock(socketID);

        if (userIndex > 0)
        {
            ref var withBlock = ref Declaraciones.UserList[userIndex];
            withBlock.incomingData.WriteBlock(ref data);

            if (withBlock.ConnID != -1) Protocol.HandleIncomingData(Convert.ToInt16(userIndex));
        }
    }

    /// <summary>
    ///     Handles closure of a socket connection
    /// </summary>
    private static void HandleConnectionClosed(int socketID)
    {
        // This runs on the main thread via the Windows.Forms event system
        var userIndex = BuscaSlotSock(socketID);

        if (userIndex > 0)
        {
            BorraSlotSock(socketID);

            Declaraciones.UserList[userIndex].ConnID = -1;
            Declaraciones.UserList[userIndex].ConnIDValida = false;

            EventoSockClose(userIndex);
        }
    }

    /// <summary>
    ///     Handles connection error
    /// </summary>
    private static void HandleConnectionError(string errorMessage)
    {
        // This runs on the main thread via the Windows.Forms event system
        Console.WriteLine("Connection Error: " + errorMessage);
    }

    /// <summary>
    ///     Closes a socket connection
    /// </summary>
    public static void Winsock_Close(int socketID)
    {
        // This is called from the main thread
        var userIndex = BuscaSlotSock(socketID);

        if (socketID > 0) SocketManager.CloseSocket(socketID);

        if (userIndex > 0)
        {
            BorraSlotSock(socketID);
            Declaraciones.UserList[userIndex].ConnID = -1;
            Declaraciones.UserList[userIndex].ConnIDValida = false;
            EventoSockClose(userIndex);
        }
    }

    /// <summary>
    ///     Initializes the socket API
    /// </summary>
    public static void IniciaWsApi(int port)
    {
        // Set up event handlers to process messages on the main thread
        SocketManager.ConnectionReceived += HandleConnectionReceived;
        SocketManager.DataReceived += HandleDataReceived;
        SocketManager.ConnectionClosed += HandleConnectionClosed;
        SocketManager.ServerError += HandleConnectionError;

        SocketManager.Initialize(port);
    }

    /// <summary>
    ///     Cleans up the socket API
    /// </summary>
    public static void LimpiaWsApi()
    {
        SocketManager.ConnectionReceived -= HandleConnectionReceived;
        SocketManager.DataReceived -= HandleDataReceived;
        SocketManager.ConnectionClosed -= HandleConnectionClosed;
        SocketManager.ServerError -= HandleConnectionError;

        SocketManager.Shutdown();

        SocketToUserMap.Clear();
    }

    /// <summary>
    ///     Sends data through a user's socket connection
    /// </summary>
    public static int WsApiEnviar(int userIndex, ref string message)
    {
        int WsApiEnviarRet = default;
        try
        {
            var returnCode = 0;

            int socketID = Declaraciones.UserList[userIndex].ConnID;

            if (socketID != -1 && Declaraciones.UserList[userIndex].ConnIDValida)
            {
                var data = SocketManager.StringToBytes(message);

                if (!SocketManager.SendData(socketID, data)) returnCode = -1;

                // Process any pending messages that may have arrived
                SocketManager.ProcessPendingMessages();
            }
            else if (socketID != -1 && !Declaraciones.UserList[userIndex].ConnIDValida)
            {
                if (!Declaraciones.UserList[userIndex].Counters.Saliendo) returnCode = -1;
            }

            WsApiEnviarRet = returnCode;
        }


        catch (Exception ex)
        {
            Console.WriteLine("Error in GetAscIP: " + ex.Message);
            Declaraciones.UserList[userIndex].outgoingData.WriteASCIIStringFixed(message);
        }

        return WsApiEnviarRet;
    }

    /// <summary>
    ///     Process any pending messages in the queue (can be called from main game loop)
    /// </summary>
    public static void ProcessNetworkMessages()
    {
        // Call this from your main game loop or timer to ensure messages get processed
        SocketManager.ProcessPendingMessages();
    }

    /// <summary>
    ///     Reinitializes all sockets
    /// </summary>
    public static void WSApiReiniciarSockets()
    {
        int i;

        var loopTo = (int)Declaraciones.MaxUsers;
        for (i = 1; i <= loopTo; i++)
            if (Declaraciones.UserList[i].ConnID != -1 && Declaraciones.UserList[i].ConnIDValida)
                TCP.CloseSocket(Convert.ToInt16(i));

        var loopTo1 = (int)Declaraciones.MaxUsers;
        for (i = 1; i <= loopTo1; i++)
        {
            Declaraciones.UserList[i].incomingData = null;
            Declaraciones.UserList[i].outgoingData = null;
        }

        Declaraciones.UserList = new Declaraciones.User[Declaraciones.MaxUsers + 1];
        ArrayInitializers.InitializeStruct(Declaraciones.UserList);
        var loopTo2 = (int)Declaraciones.MaxUsers;
        for (i = 1; i <= loopTo2; i++)
        {
            Declaraciones.UserList[i].ConnID = -1;
            Declaraciones.UserList[i].ConnIDValida = false;

            Declaraciones.UserList[i].incomingData = new clsByteQueue();
            Declaraciones.UserList[i].outgoingData = new clsByteQueue();
        }

        Declaraciones.LastUser = 1;
        Declaraciones.NumUsers = 0;

        SocketToUserMap.Clear();

        IniciaWsApi(Admin.Puerto);
    }

    /// <summary>
    ///     Finds the user index associated with a socket ID
    /// </summary>
    public static int BuscaSlotSock(int socketID)
    {
        var userIndex = -1;

        if (SocketToUserMap.TryGetValue(socketID, out userIndex)) return userIndex;

        return -1;
    }

    /// <summary>
    ///     Associates a socket ID with a user index
    /// </summary>
    public static void AgregaSlotSock(int socketID, int userIndex)
    {
        Console.WriteLine("AgregaSlotSock: Socket " + socketID + " -> User " + userIndex);

        if (SocketToUserMap.Count > Declaraciones.MaxUsers)
        {
            TCP.CloseSocket(Convert.ToInt16(userIndex));
            return;
        }

        SocketToUserMap[socketID] = userIndex;
    }

    /// <summary>
    ///     Removes the association between a socket ID and a user index
    /// </summary>
    public static void BorraSlotSock(int socketID)
    {
        var count = SocketToUserMap.Count;

        SocketToUserMap.Remove(socketID);

        Console.WriteLine("BorraSlotSock: " + count + " -> " + SocketToUserMap.Count);
    }

    /// <summary>
    ///     Handles the event of a socket being closed
    /// </summary>
    public static void EventoSockClose(int userIndex)
    {
        if (modCentinela.Centinela.RevisandoUserIndex == userIndex) modCentinela.CentinelaUserLogout();

        if (Declaraciones.UserList[userIndex].flags.UserLogged)
        {
            TCP.CloseSocketSL(Convert.ToInt16(userIndex));
            UsUaRiOs.Cerrar_Usuario(Convert.ToInt16(userIndex));
        }
        else
        {
            TCP.CloseSocket(Convert.ToInt16(userIndex));
        }
    }
}